File: src/providers/codex.py
"""Codex provider 实现。"""

from __future__ import annotations

import json
import os
import tempfile
from pathlib import Path
from typing import Any, Dict, List, Optional

from src.utils.process_runner import (
    ProcessExecutionError,
    resolve_timeout_config,
    run_stream_process,
)


def _list_directories(path: Path) -> List[str]:
    """安全获取目录下一级子目录名称。"""
    try:
        return [item.name for item in path.iterdir() if item.is_dir()]
    except Exception:
        return []


def _find_vscode_bundled_codex() -> Optional[str]:
    """Windows 下尝试定位 VS Code 扩展内置 codex.exe。"""
    if os.name != "nt":
        return None

    user_profile = os.environ.get("USERPROFILE")
    if not user_profile:
        return None

    extensions_dir = Path(user_profile) / ".vscode" / "extensions"
    extension_dirs = sorted(
        (
            name
            for name in _list_directories(extensions_dir)
            if name.startswith("openai.chatgpt-")
        ),
        reverse=True,
    )

    for extension_dir in extension_dirs:
        exe_path = (
            extensions_dir
            / extension_dir
            / "bin"
            / "windows-x86_64"
            / "codex.exe"
        )
        if exe_path.exists():
            return str(exe_path)

    return None


def resolve_codex_command() -> str:
    """解析最终要执行的 codex 命令路径。"""
    if os.environ.get("CODEX_BIN"):
        return os.environ["CODEX_BIN"]

    bundled = _find_vscode_bundled_codex()
    if bundled:
        return bundled

    return "codex"


def _text_from_parts(value: Any) -> str:
    """从多种消息结构中抽取纯文本。"""
    if isinstance(value, str):
        return value
    if value is None:
        return ""
    if isinstance(value, list):
        return "".join(_text_from_parts(item) for item in value)
    if isinstance(value, dict):
        if isinstance(value.get("text"), str):
            return value["text"]
        if isinstance(value.get("output_text"), str):
            return value["output_text"]
        if isinstance(value.get("content"), list):
            return _text_from_parts(value["content"])
        if value.get("delta") is not None:
            return _text_from_parts(value["delta"])
        if value.get("message") is not None:
            return _text_from_parts(value["message"])
    return ""


def _extract_assistant_text(event: Dict[str, Any], state: Dict[str, Any]) -> str:
    """从单条 Codex JSON 事件中抽取助手输出文本。"""
    if not isinstance(event, dict):
        return ""

    if event.get("type") == "thread.started" and isinstance(event.get("thread_id"), str):
        state["thread_id"] = event["thread_id"]

    if event.get("type") == "item.completed" and isinstance(event.get("item"), dict):
        item = event["item"]
        if item.get("type") in {"agent_message", "assistant"}:
            if state["saw_delta"]:
                return ""
            return _text_from_parts(item.get("text") or item.get("message") or item.get("content"))

    if event.get("type") == "agent_message_delta":
        state["saw_delta"] = True
        return _text_from_parts(event.get("delta"))

    if event.get("type") == "agent_message":
        if state["saw_delta"]:
            return ""
        return _text_from_parts(event.get("message"))

    if event.get("type") == "assistant":
        if state["saw_delta"]:
            return ""
        return _text_from_parts(event.get("message") or event.get("content"))

    if event.get("role") == "assistant":
        if state["saw_delta"]:
            return ""
        return _text_from_parts(event.get("content") or event.get("message") or event.get("delta"))

    return ""


def invoke_codex(
    prompt: str,
    session_id: Optional[str] = None,
    stream: bool = True,
    *,
    workdir: Optional[str] = None,
    exec_mode: str = "safe",
    output_schema: Optional[Any] = None,
    timeout_level: str = "standard",
    idle_timeout_s: Optional[float] = None,
    max_timeout_s: Optional[float] = None,
    terminate_grace_s: Optional[float] = None,
) -> Dict[str, Any]:
    """调用 Codex CLI 并解析流式输出。"""
    if not isinstance(prompt, str) or not prompt.strip():
        raise ValueError("prompt must be a non-empty string")

    codex_command = resolve_codex_command()
    base_flags = ["--json", "--skip-git-repo-check"]
    exec_prefix = ["exec"]
    normalized_exec_mode = (exec_mode or "safe").strip().lower()
    if normalized_exec_mode == "full_auto":
        exec_prefix.append("--full-auto")
    elif normalized_exec_mode == "bypass":
        exec_prefix.append("--dangerously-bypass-approvals-and-sandbox")
    elif normalized_exec_mode != "safe":
        raise ValueError("exec_mode must be one of: safe, full_auto, bypass")

    args = (
        [*exec_prefix, "resume", *base_flags, session_id, prompt]
        if session_id
        else [*exec_prefix, *base_flags, prompt]
    )

    schema_temp_file: Optional[str] = None
    if output_schema is not None:
        schema_path = ""
        if isinstance(output_schema, str):
            candidate = Path(output_schema)
            if candidate.exists():
                schema_path = str(candidate)
            else:
                schema_path = output_schema
        else:
            with tempfile.NamedTemporaryFile(
                mode="w", suffix=".json", encoding="utf-8", delete=False
            ) as fp:
                json.dump(output_schema, fp, ensure_ascii=False, indent=2)
                schema_temp_file = fp.name
                schema_path = schema_temp_file
        args += ["--output-schema", schema_path]

    state: Dict[str, Any] = {
        "saw_delta": False,
        "thread_id": session_id,
        "printed_any": False,
        "needs_newline": False,
        "output_parts": [],
    }

    def on_stdout_line(line: str) -> None:
        trimmed = line.strip()
        if not trimmed:
            return
        try:
            event = json.loads(trimmed)
        except json.JSONDecodeError:
            return

        text = _extract_assistant_text(event, state)
        if not text:
            return

        state["output_parts"].append(text)
        if stream:
            print(text, end="", flush=True)
            state["printed_any"] = True
            state["needs_newline"] = not text.endswith("\n")

    timeout = resolve_timeout_config(
        timeout_level=timeout_level,
        idle_timeout_s=idle_timeout_s,
        max_timeout_s=max_timeout_s,
        terminate_grace_s=terminate_grace_s,
    )

    try:
        try:
            result = run_stream_process(
                provider="codex",
                command=codex_command,
                args=args,
                workdir=workdir,
                timeout=timeout,
                stream_stderr=stream,
                stderr_prefix="[codex stderr] ",
                on_stdout_line=on_stdout_line,
            )
        except ProcessExecutionError as exc:
            raise ProcessExecutionError(
                provider=exc.provider,
                reason=exc.reason,
                command_repr=exc.command_repr,
                elapsed_ms=exc.elapsed_ms,
                return_code=exc.return_code,
                stderr_lines=exc.stderr_lines,
                session_id=state.get("thread_id"),
                extra_message=exc.extra_message,
            ) from exc
    finally:
        if schema_temp_file:
            try:
                Path(schema_temp_file).unlink(missing_ok=True)
            except Exception:
                pass

    if stream and state["printed_any"] and state["needs_newline"]:
        print("")

    return {
        "provider": "codex",
        "text": "".join(state["output_parts"]),
        "session_id": state.get("thread_id"),
        "elapsed_ms": result.elapsed_ms,
    }

File: src/providers/codex.js
"use strict";

const { spawn } = require("child_process");
const fs = require("fs");
const path = require("path");
const readline = require("readline");

// 安全列出目录下的子目录名称。
function listDirectories(dir) {
  try {
    return fs
      .readdirSync(dir, { withFileTypes: true })
      .filter((entry) => entry.isDirectory())
      .map((entry) => entry.name);
  } catch {
    return [];
  }
}

// Windows 场景：尝试定位 VS Code 扩展里自带的 codex.exe。
// 这样即使 PATH 里没有 codex，也能尽量运行起来。
function findVscodeBundledCodex() {
  if (process.platform !== "win32") return null;

  const userProfile = process.env.USERPROFILE;
  if (!userProfile) return null;

  const extensionsDir = path.join(userProfile, ".vscode", "extensions");
  const extensionDirs = listDirectories(extensionsDir)
    .filter((name) => name.startsWith("openai.chatgpt-"))
    .sort((a, b) => b.localeCompare(a, undefined, { numeric: true }));

  for (const extensionDir of extensionDirs) {
    const exePath = path.join(
      extensionsDir,
      extensionDir,
      "bin",
      "windows-x86_64",
      "codex.exe"
    );
    if (fs.existsSync(exePath)) return exePath;
  }

  return null;
}

// 决定要启动的 codex 命令。
// 优先级：CODEX_BIN > VS Code 内置 codex.exe > PATH 中的 codex。
function resolveCodexCommand() {
  if (process.env.CODEX_BIN) return process.env.CODEX_BIN;

  const bundled = findVscodeBundledCodex();
  if (bundled) return bundled;

  return "codex";
}

// 从可能的嵌套结构中提取纯文本。
function textFromParts(value) {
  if (typeof value === "string") return value;
  if (!value) return "";

  if (Array.isArray(value)) return value.map(textFromParts).join("");

  if (typeof value === "object") {
    if (typeof value.text === "string") return value.text;
    if (typeof value.output_text === "string") return value.output_text;
    if (Array.isArray(value.content)) return textFromParts(value.content);
    if (value.delta) return textFromParts(value.delta);
    if (value.message) return textFromParts(value.message);
  }

  return "";
}

// 兼容多种事件格式，统一抽取“助手文本”。
function extractAssistantText(event, state) {
  if (!event || typeof event !== "object") return "";

  // 记录/更新线程 id，用于后续 resume。
  if (event.type === "thread.started" && typeof event.thread_id === "string") {
    state.threadId = event.thread_id;
  }

  if (event.type === "item.completed" && event.item) {
    const itemType = event.item.type;
    if (itemType === "agent_message" || itemType === "assistant") {
      if (state.sawDelta) return "";
      return textFromParts(event.item.text || event.item.message || event.item.content);
    }
  }

  if (event.type === "agent_message_delta") {
    state.sawDelta = true;
    return textFromParts(event.delta);
  }

  if (event.type === "agent_message") {
    if (state.sawDelta) return "";
    return textFromParts(event.message);
  }

  if (event.type === "assistant") {
    if (state.sawDelta) return "";
    return textFromParts(event.message || event.content);
  }

  if (event.role === "assistant") {
    if (state.sawDelta) return "";
    return textFromParts(event.content || event.message || event.delta);
  }

  return "";
}

// 调用 codex 并流式返回文本。
// options:
// - prompt: 当前要发送的问题
// - sessionId: 若存在则使用 exec resume 继续会话
// - stream: 是否实时打印输出（默认 true）
function invokeCodex(options) {
  const { prompt, sessionId = null, stream = true } = options;

  return new Promise((resolve, reject) => {
    const codexCommand = resolveCodexCommand();
    const baseFlags = ["--json", "--skip-git-repo-check"];

    const args = sessionId
      ? ["exec", "resume", ...baseFlags, sessionId, prompt]
      : ["exec", ...baseFlags, prompt];

    const child = spawn(codexCommand, args, {
      stdio: ["ignore", "pipe", "pipe"],
    });

    const state = {
      sawDelta: false,
      threadId: sessionId,
      printedAny: false,
      needsNewline: false,
      output: "",
      stderr: [],
    };

    const stdoutRl = readline.createInterface({
      input: child.stdout,
      crlfDelay: Infinity,
    });

    stdoutRl.on("line", (line) => {
      const trimmed = line.trim();
      if (!trimmed) return;

      let event;
      try {
        event = JSON.parse(trimmed);
      } catch {
        return;
      }

      const text = extractAssistantText(event, state);
      if (!text) return;

      state.output += text;

      if (stream) {
        process.stdout.write(text);
        state.printedAny = true;
        state.needsNewline = !text.endsWith("\n");
      }
    });

    const stderrRl = readline.createInterface({
      input: child.stderr,
      crlfDelay: Infinity,
    });

    stderrRl.on("line", (line) => {
      if (!line.trim()) return;
      state.stderr.push(line);
      if (stream) {
        console.error(`[codex stderr] ${line}`);
      }
    });

    child.on("error", (err) => {
      if (err.code === "ENOENT") {
        reject(
          new Error(
            `Failed to start codex: ${err.message}. Tried command: ${codexCommand}.`
          )
        );
        return;
      }
      reject(new Error(`Failed to start codex: ${err.message}`));
    });

    child.on("close", (code, signal) => {
      if (stream && state.printedAny && state.needsNewline) {
        process.stdout.write("\n");
      }

      if (signal) {
        reject(new Error(`codex terminated by signal: ${signal}`));
        return;
      }

      if (code !== 0) {
        reject(
          new Error(
            `codex exited with code ${code}${
              state.stderr.length ? `: ${state.stderr.join(" | ")}` : ""
            }`
          )
        );
        return;
      }

      resolve({
        provider: "codex",
        text: state.output,
        sessionId: state.threadId || null,
      });
    });
  });
}

module.exports = {
  invokeCodex,
};

